File: dertet-harness-android/app/src/main/java/com/dertet/harness/network/GeminiClient.kt
package com.dertet.harness.network

import com.dertet.harness.data.AttachmentKind
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.buildJsonArray
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.put
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody

class GeminiClient : ChatClient {

    override fun streamChat(
        baseUrl: String,
        apiKey: String,
        model: String,
        systemPrompt: String,
        history: List<ChatMessage>
    ): Flow<ChatStreamEvent> = flow {
        val contents = buildJsonArray {
            for (msg in history) {
                val role = if (msg.role == "assistant") "model" else "user"
                add(buildJsonObject {
                    put("role", role)
                    put("parts", buildParts(msg))
                })
            }
        }

        val bodyBuilder = buildJsonObject {
            put("contents", contents)
            if (systemPrompt.isNotBlank()) {
                put("systemInstruction", buildJsonObject {
                    put("parts", buildJsonArray {
                        add(buildJsonObject { put("text", systemPrompt) })
                    })
                })
            }
        }
        val body = bodyBuilder.toString().toRequestBody("application/json".toMediaType())

        val cleanModel = model.removePrefix("models/")
        val url = baseUrl.trimEnd('/') + "/models/$cleanModel:streamGenerateContent?alt=sse"
        val request = Request.Builder()
            .url(url)
            .addHeader("x-goog-api-key", apiKey)
            .addHeader("Content-Type", "application/json")
            .post(body)
            .build()

        var gotAny = false
        try {
            sseFlow(sharedHttpClient, request).collect { event ->
                val data = event.data.trim()
                if (data.isEmpty()) return@collect
                try {
                    val json = Json.parseToJsonElement(data).jsonObject
                    val errorObj = json["error"]?.jsonObject
                    if (errorObj != null) {
                        val msg = errorObj["message"]?.jsonPrimitive?.contentOrNull ?: "API error"
                        emit(ChatStreamEvent.Error(msg))
                        return@collect
                    }
                    val candidates = json["candidates"]?.jsonArray
                    val parts = candidates?.firstOrNull()?.jsonObject?.get("content")?.jsonObject
                        ?.get("parts")?.jsonArray
                    val text = parts?.joinToString("") { it.jsonObject["text"]?.jsonPrimitive?.contentOrNull ?: "" }
                    if (!text.isNullOrEmpty()) {
                        gotAny = true
                        emit(ChatStreamEvent.Delta(text))
                    }
                    parts?.forEach { part ->
                        val inline = part.jsonObject["inlineData"]?.jsonObject
                            ?: part.jsonObject["inline_data"]?.jsonObject
                        val mime = inline?.get("mimeType")?.jsonPrimitive?.contentOrNull
                            ?: inline?.get("mime_type")?.jsonPrimitive?.contentOrNull
                        val imgData = inline?.get("data")?.jsonPrimitive?.contentOrNull
                        if (!mime.isNullOrEmpty() && !imgData.isNullOrEmpty()) {
                            gotAny = true
                            emit(ChatStreamEvent.ImageGenerated(mime, imgData))
                        }
                    }
                } catch (e: Exception) {
                    // ignore malformed chunk
                }
            }
            emit(ChatStreamEvent.Done)
        } catch (e: Exception) {
            if (!gotAny) emit(ChatStreamEvent.Error(e.message ?: "Помилка з’єднання"))
            else emit(ChatStreamEvent.Done)
        }
    }

    private fun buildParts(msg: ChatMessage): kotlinx.serialization.json.JsonElement {
        val images = msg.attachments.filter { it.kind == AttachmentKind.IMAGE && it.base64Data != null }
        val files = msg.attachments.filter { it.kind == AttachmentKind.FILE }

        var textPart = msg.text
        for (f in files) {
            if (f.textContent != null) {
                textPart = "[Файл: ${f.fileName}]\n${f.textContent}\n\n$textPart"
            }
        }

        return buildJsonArray {
            add(buildJsonObject { put("text", textPart) })
            for (img in images) {
                add(buildJsonObject {
                    put("inline_data", buildJsonObject {
                        put("mime_type", img.mimeType)
                        put("data", img.base64Data)
                    })
                })
            }
        }
    }
}

File: dertet-harness-ios/DertetHarness/Network/GeminiClient.swift
import Foundation

final class GeminiClient: ChatClient {
    func streamChat(
        baseUrl: String,
        apiKey: String,
        model: String,
        systemPrompt: String,
        history: [ChatMessage]
    ) -> AsyncThrowingStream<ChatStreamEvent, Error> {
        AsyncThrowingStream { continuation in
            let task = Task {
                var gotAny = false
                do {
                    var contents: [[String: Any]] = []
                    for msg in history {
                        let role = msg.role == "assistant" ? "model" : "user"
                        contents.append(["role": role, "parts": buildParts(msg)])
                    }

                    var body: [String: Any] = ["contents": contents]
                    if !systemPrompt.isEmpty {
                        body["systemInstruction"] = ["parts": [["text": systemPrompt]]]
                    }

                    let cleanModel = model.hasPrefix("models/") ? String(model.dropFirst("models/".count)) : model
                    guard let url = URL(string: baseUrl.trimmedTrailingSlash + "/models/\(cleanModel):streamGenerateContent?alt=sse") else {
                        continuation.finish(throwing: URLError(.badURL))
                        return
                    }
                    var request = URLRequest(url: url)
                    request.httpMethod = "POST"
                    request.setValue("application/json", forHTTPHeaderField: "Content-Type")
                    request.setValue(apiKey, forHTTPHeaderField: "x-goog-api-key")
                    request.httpBody = jsonString(body)

                    for try await event in sseEvents(for: request) {
                        try Task.checkCancellation()
                        let data = event.data.trimmingCharacters(in: .whitespaces)
                        if data.isEmpty { continue }
                        guard let jsonData = data.data(using: .utf8),
                              let obj = try? JSONSerialization.jsonObject(with: jsonData) as? [String: Any] else { continue }

                        if let errorObj = asDict(obj["error"]), let msg = asString(errorObj["message"]) {
                            continuation.yield(.error(msg))
                            continue
                        }
                        if let candidates = asArray(obj["candidates"]),
                           let first = asDict(candidates.first),
                           let content = asDict(first["content"]),
                           let parts = asArray(content["parts"]) {
                            let text = parts.compactMap { asString(asDict($0)?["text"]) }.joined()
                            if !text.isEmpty {
                                gotAny = true
                                continuation.yield(.delta(text))
                            }
                        }
                    }
                    continuation.yield(.done)
                    continuation.finish()
                } catch {
                    if !gotAny {
                        continuation.yield(.error(error.localizedDescription))
                    }
                    continuation.finish()
                }
            }
            continuation.onTermination = { _ in task.cancel() }
        }
    }

    private func buildParts(_ msg: ChatMessage) -> [[String: Any]] {
        let images = msg.attachments.filter { $0.kind == .image && $0.base64Data != nil }
        let files = msg.attachments.filter { $0.kind == .file }

        var text = msg.text
        for f in files where f.textContent != nil {
            text = "[File: \(f.fileName)]\n\(f.textContent!)\n\n\(text)"
        }

        var parts: [[String: Any]] = [["text": text]]
        for img in images {
            parts.append(["inline_data": ["mime_type": img.mimeType, "data": img.base64Data!]])
        }
        return parts
    }
}
